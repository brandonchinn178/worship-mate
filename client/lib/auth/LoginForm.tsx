import { useForm } from 'react-hook-form'
import { toast } from 'react-toastify'
import styled from 'styled-components'

import { font, fontFamily } from '~/theme'

export type LoginFormValues = {
  username: string
  password: string
}

export type LoginFormProps = {
  onSubmit: (payload: LoginFormValues) => Promise<void>
}

export function LoginForm({ onSubmit }: LoginFormProps) {
  const {
    formState,
    handleSubmit,
    register,
    setError,
  } = useForm<LoginFormValues>()

  const { isSubmitting, isSubmitSuccessful } = formState
  const postLoginLoading = isSubmitting || isSubmitSuccessful

  const onFormSubmit = handleSubmit(async (payload) => {
    try {
      await onSubmit(payload)
    } catch (e) {
      toast.error(e.toString())
      console.error(e)

      // tell react-hook-form that the submission was not successful
      // https://github.com/react-hook-form/react-hook-form/issues/3956
      setError('password', { message: e.message })
    }
  })

  return (
    <LoginFormContainer onSubmit={onFormSubmit}>
      <LoginLabel htmlFor="username">Username</LoginLabel>
      <LoginInput
        id="username"
        name="username"
        ref={register({ required: true })}
        autoFocus
      />
      <LoginLabel htmlFor="password">Password</LoginLabel>
      <LoginInput
        id="password"
        name="password"
        type="password"
        ref={register({ required: true })}
      />
      <LoginButton disabled={postLoginLoading}>Login</LoginButton>
    </LoginFormContainer>
  )
}

const LoginFormContainer = styled.form`
  display: grid;
  grid-columns: 1fr;

  margin: 0 auto;
  max-width: 500px;

  padding: 25px 50px;
`

const LoginLabel = styled.label`
  margin-top: 5px;

  ${font('label')}
`

const LoginInput = styled.input`
  margin-bottom: 0.5rem;
  padding: 2px 5px;

  font-family: ${fontFamily('notoSerif')};
`

const LoginButton = styled.button`
  margin: 20px auto 0;
  padding: 5px 20px;
  width: min-content;
`
