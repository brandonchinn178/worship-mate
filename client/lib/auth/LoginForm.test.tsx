import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderUI, waitForAllStateChanges } from '~jest-utils'

import { LoginForm } from './LoginForm'

/** Mock react-toastify **/

const mockToastError = jest.fn()
jest.mock('react-toastify', () => {
  return {
    toast: {
      error: (...args: unknown[]) => mockToastError(...args),
    },
  }
})

/** Mock console.error **/

afterEach(jest.restoreAllMocks)

const mockConsoleError = () =>
  jest.spyOn(console, 'error').mockImplementation(() => {})

/** Tests **/

it('renders the login form', () => {
  const { container } = renderUI(<LoginForm onSubmit={jest.fn()} />)

  expect(container).toMatchSnapshot()
  expect(screen.getByLabelText('Username')).toBeVisible()
  expect(screen.getByLabelText('Password')).toBeVisible()
  expect(screen.getByText('Login', { selector: 'button' })).toBeVisible()
})

it('shows errors when onSubmit fails', async () => {
  const e = new Error('onSubmit failed')
  const onSubmit = jest.fn().mockRejectedValue(e)
  const consoleError = mockConsoleError()

  renderUI(<LoginForm onSubmit={onSubmit} />)

  userEvent.type(screen.getByLabelText('Username'), 'user')
  userEvent.type(screen.getByLabelText('Password'), 'pw')
  userEvent.click(screen.getByText('Login'))

  await waitFor(() => {
    expect(onSubmit).toHaveBeenCalled()
  })

  expect(consoleError).toHaveBeenCalledWith(e)
  expect(mockToastError).toHaveBeenCalledWith(
    expect.stringContaining('onSubmit failed'),
  )
})

describe('form validation', () => {
  it('triggers onSubmit when passes', async () => {
    const onSubmit = jest.fn()
    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.type(screen.getByLabelText('Username'), 'user')
    userEvent.type(screen.getByLabelText('Password'), 'pw')
    userEvent.click(screen.getByText('Login'))

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })
  })

  it('does not trigger onSubmit when username is missing', async () => {
    const onSubmit = jest.fn()
    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.type(screen.getByLabelText('Password'), 'pw')
    userEvent.click(screen.getByText('Login'))

    await waitForAllStateChanges()
    expect(onSubmit).not.toHaveBeenCalled()
  })

  it('does not trigger onSubmit when password is missing', async () => {
    const onSubmit = jest.fn()
    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.type(screen.getByLabelText('Username'), 'user')
    userEvent.click(screen.getByText('Login'))

    await waitForAllStateChanges()
    expect(onSubmit).not.toHaveBeenCalled()
  })

  it('does not trigger onSubmit when both username and password are missing', async () => {
    const onSubmit = jest.fn()

    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.click(screen.getByText('Login'))

    await waitForAllStateChanges()
    expect(onSubmit).not.toHaveBeenCalled()
  })
})

describe('submit button', () => {
  it('is enabled at start', async () => {
    renderUI(<LoginForm onSubmit={jest.fn()} />)

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })

  it('is disabled during and after submitting', async () => {
    const onSubmit = jest.fn()
    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.type(screen.getByLabelText('Username'), 'user')
    userEvent.type(screen.getByLabelText('Password'), 'pw')
    userEvent.click(screen.getByText('Login'))

    expect(screen.getByRole('button')).toHaveAttribute('disabled')

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })

    expect(screen.getByRole('button')).toHaveAttribute('disabled')
  })

  it('is enabled after form validation fails', async () => {
    const onSubmit = jest.fn()
    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.click(screen.getByText('Login'))

    await waitForAllStateChanges()
    expect(onSubmit).not.toHaveBeenCalled()

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })

  it('is enabled after submission fails', async () => {
    const onSubmit = jest.fn().mockRejectedValue(new Error())
    mockConsoleError()

    renderUI(<LoginForm onSubmit={onSubmit} />)

    userEvent.type(screen.getByLabelText('Username'), 'user')
    userEvent.type(screen.getByLabelText('Password'), 'pw')
    userEvent.click(screen.getByText('Login'))

    expect(screen.getByRole('button')).toHaveAttribute('disabled')

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalled()
    })

    expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
  })
})
