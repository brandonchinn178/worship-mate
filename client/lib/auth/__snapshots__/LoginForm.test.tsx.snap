// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the login form 1`] = `
<div>
  <form
    class="LoginForm__LoginFormContainer-vdxufc-0 ehysyc"
  >
    <label
      class="LoginForm__LoginLabel-vdxufc-1 eVfdSf"
      for="username"
    >
      Username
    </label>
    <input
      class="LoginForm__LoginInput-vdxufc-2 kqWixQ"
      id="username"
      name="username"
    />
    <label
      class="LoginForm__LoginLabel-vdxufc-1 eVfdSf"
      for="password"
    >
      Password
    </label>
    <input
      class="LoginForm__LoginInput-vdxufc-2 kqWixQ"
      id="password"
      name="password"
      type="password"
    />
    <button
      class="LoginForm__LoginButton-vdxufc-3 gLgGik"
    >
      Login
    </button>
  </form>
</div>
`;
